wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java

```diff
@@ -110,13 +135,97 @@ protected boolean doHandleRequest(final Request request) {
 		boolean changed = value != null || current != null;
 
 		if (changed) {
-			setData(value);
-		}
+			// Reset validation fields
+			resetValidationState();
+			
+			// if fileType is supplied then validate it
+			boolean validFileType;
+			if (hasFileTypes()) {
+				validFileType = FileUtil.validateFileType(value, getFileTypes());
+				getOrCreateComponentModel().validFileType = validFileType;
+			} else {
+				validFileType = true;
+			}
+			
+			// if fileSize is supplied then validate it
+			boolean validFileSize;
+			if (hasMaxFileSize()) {
+				validFileSize = FileUtil.validateFileSize(value, getMaxFileSize());
+				getOrCreateComponentModel().validFileSize = validFileSize;
+			} else {
+				validFileSize = true;
+			}
+			
+			// if file is valid, the update data
+			if (validFileSize && validFileType) {
+				setData(value);
+			} else if (current == null) {
+				// otherwise no change
+				changed = false;
+			} else {
+				changed = true;
+				setData(null);
+			}
+		} 
 
 		return changed;
 	}
 
 	/**
+	 * Reset validation state.
+	 */
+	public void resetValidationState() {
+		getOrCreateComponentModel().validFileType = null;
+		getOrCreateComponentModel().validFileSize = null;
+	}
+
+	/**
+	 * Indicates whether the uploaded file is valid, if {@link getFileTypes()} is set.
+	 *
+	 * @return true if file type valid, false file type invalid, otherwise null.
+	 * @see {@link getFileValidationMessages()}
+	 */
+	public Boolean isFileTypeValid() {
+		return getComponentModel().validFileType;
+	}
+
+	/**
+	 * Indicates whether the uploaded file is valid, if {@link getMaxFileSize()} is set.
+	 *
+	 * @return true if file size valid, false file size invalid, otherwise null.
+	 * @see {@link getFileValidationMessages()}
+	 */
+	public Boolean isFileSizeValid() {
+		return getComponentModel().validFileSize;
+	}
+	
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void validateComponent(final List<Diagnostic> diags) {
+		super.validateComponent(diags);
+		
```

Comment:
This validation logic looks good. 
What do you think of putting the `create error message` in the `FileUtil` so it can be shared between both file widgets.